py/usageaccountant/datadog_fetcher.py

```diff
@@ -0,0 +1,307 @@
+             "scope":"app_feature: shared, shared_resource_id: rc_long_redis",
+             "interval":5,
+             "length":2,
+             "start":1721083885000,
+             "end":1721083894000,
+             "pointlist":[
+                [1721083885000.0, 2494386918.6666665],
+                [1721083890000.0, 2494386918.6666665]
+             ],
+             "display_name":"redis.mem.peak",
+             "attributes":{}
+          }
+       ],
+       "values":[],
+       "times":[],
+       "message":"",
+       "group_by":["shared_resource_id"]
+    }
+
+    series_list: List of series dict extracted from API response
+    parsed_unit: UsageUnit parsed from API response
+    usage_accumulated: Object of UsageAccumulator class
+    """
+    for series in series_list:
+        scope = series.get("scope", "")
+        scope_dict = parse_response_scope(scope)
+        if (
+            "shared_resource_id" not in scope_dict.keys()
+            or "app_feature" not in scope_dict.keys()
+        ):
+            exception_msg = (
+                "Required parameters, shared_resource_id and/or app_feature "
+                "not found in series.scope of the series received."
+            )
+            raise InvalidSeriesError(exception_msg, series)
+
+        resource = scope_dict["shared_resource_id"]
+        app_feature = scope_dict["app_feature"]
+        for point in series.get("pointlist", []):
+            # TODO remove this before prod
+            # print(f"resource: {resource}, app_feature: {app_feature}, "
+            #       f"amount: {int(point[1])}, usage_type: {parsed_unit}")
+            usage_accumulator.record(
+                resource_id=resource,
+                app_feature=app_feature,
+                # point[0] is the timestamp
+                amount=int(point[1]),
+                usage_type=parsed_unit,
+            )
+
+
+def main(
+    query_file: TextIO,
+    start_time: int,
+    end_time: int,
+    kafka_config_file: TextIO,
+) -> None:
+    """
+    query: File with Datadog queries, one per line
+    start_time: Start of the time window in Unix epoch format
+                with second-level precision
+    end_time: End of the time window in Unix epoch format
+              with second-level precision
+    kafka_config: File with parameters to initialize UsageAccumulator object
+    """
+    kafka_config = json.loads(kafka_config_file.read())
+    ua = UsageAccumulator(kafka_config=kafka_config)
+
+    for line in query_file:
+        query = line.rstrip()
+        if not is_valid_query(query):
+            raise ValueError(
+                f"Required parameter shared_resource_id and/or "
+                f"app_feature not found in the query: {query}."
+            )
+
+        response = query_datadog(query, start_time, end_time)
+
+        series_list, parsed_unit = parse_response_series_and_unit(response)
+        parse_and_post(series_list, parsed_unit, ua)
```

Comment:
This would make the job non idempotent. If the job fails half way through and is retried, some queries will be recorded twice, other not, generating potentially wrong results. Imagine if it starts crashlooping.

You should collect all the usage data first, then send them to the Accountant. so A failure on a query would make the whole job retry but you will not have produced anything on kafka at that point.